### PR TITLE
[Model runner V2][Performance] optimize gumbel sample kernel

### LINKED ISSUE #7608: [Performance] AscendSampler vs v2 Gumbel Sampling: ~55× latency gap on NPU
## Summary

Benchmark shows that the **v1 AscendSampler** (`vllm_ascend/sample/sampler.py`) sampling path is **~56× faster** than the **v2 Gumbel Triton kernel** (`vllm_ascend/worker/v2/sample/gumbel.py`) under the same workload. Both paths implement Gumbel-max categorical sampling; the gap suggests optimization opportunity for v2 model runner.

## Environment

- **Device**: Ascend NPU
- **Timing**: `torch.npu.Event` (NPU-side only)

## Two Sampling Paths

| Path | Location | Implementation |
|------|----------|----------------|
| **AscendSampler (v1)** | `vllm_ascend/sample/sampler.py` | `softmax` → `exponential` → `probs.div_(q)` → `argmax` (PyTorch) |
| **v2 Gumbel** | `vllm_ascend/worker/v2/sample/gumbel.py` | Triton kernel: add Gumbel noise to logits block-wise, then argmax per block and across blocks |

Both are mathematically equivalent (Gumbel-max trick for categorical sampling).

## Benchmark Results

**Config**: `vocab_size=128000`, `temperature=0.8`, `num_reqs ∈ {1, 8, 32, 128}`, warmup=5, iters=50

| num_reqs | AscendSampler (sample/sampler.py) | v2 gumbel.py |
|----------|-----------------------------------|--------------|
| 1 | 0.54 ms | 1.40 ms |
| 8 | 0.56 ms | 7.90 ms |
| 32 | 0.89 ms | 31.12 ms |
| 128 | 2.25 ms | 124.00 ms |

**Speedup (AscendSampler / v2 gumbel)**: ~2.6× (num_reqs=1) → **~55×** (num_reqs=128)

## Reproducibility

Copy the following script to a `.py` file and run it in the vllm-ascend repo (with NPU available):

```python
#!/usr/bin/env python3
"""AscendSampler vs v2 gumbel performance comparison.
Uses AscendSampler instance and forward().
"""
import torch

from vllm.config import VllmConfig
from vllm.v1.sample.logits_processor import LogitsProcessors
from vllm.v1.sample.metadata import SamplingMetadata
from vllm_ascend.ascend_config import init_ascend_config
from vllm_ascend.ops.triton.triton_utils import init_device_properties_triton
from vllm_ascend.sample.sampler import AscendSampler
from vllm_ascend.worker.v2.sample.gumbel import gumbel_sample

VOCAB_SIZE = 128000
NUM_REQS_LIST = [1, 8, 32, 128]
NUM_WARMUP = 5
NUM_ITERS = 50


def make_sampling_metadata(num_reqs: int, temperature: torch.Tensor, device: str):
    return SamplingMetadata(
        temperature=temperature,
        all_greedy=False,
        all_random=True,
        top_p=None,
        top_k=None,
        generators={},
        max_num_logprobs=None,
        no_penalties=True,
        prompt_token_ids=None,
        frequency_penalties=torch.zeros(num_reqs, dtype=torch.float32, device=device),
        presence_penalties=torch.zeros(num_reqs, dtype=torch.float32, device=device),
        repetition_penalties=torch.ones(num_reqs, dtype=torch.float32, device=device),
        output_token_ids=[[] for _ in range(num_reqs)],
        allowed_token_ids_mask=None,
        bad_words_token_ids={},
        logitsprocs=LogitsProcessors(),
    )


def main():
    if not torch.npu.is_available():
        print("SKIP: NPU not available")
        return

    init_device_properties_triton()
    init_ascend_config(VllmConfig())
    device = "npu"
    torch.manual_seed(42)

    start_evt = torch.npu.Event(enable_timing=True)
    end_evt = torch.npu.Event(enable_timing=True)
    ascend_sampler = AscendSampler()

    print("AscendSampler vs v2 gumbel performance comparison")
    print("=" * 60)
    print(f"vocab_size={VOCAB_SIZE}, warmup={NUM_WARMUP}, iters={NUM_ITERS}")
    print()

    for num_reqs in NUM_REQS_LIST:
        logits = torch.randn(num_reqs, VOCAB_SIZE, dtype=torch.float32, device=device)
        temperature = torch.ones(num_reqs, dtype=torch.float32, device=device) * 0.8
        idx_mapping = torch.arange(num_reqs, dtype=torch.int64, device=device)
        seed = torch.randint(0, 2**31, (num_reqs,), dtype=torch.int64, device=device)
        pos = torch.randint(0, 100, (num_reqs,), dtype=torch.int64, device=device)
        sampling_metadata = make_sampling_metadata(num_reqs, temperature, device)

        def run_ascend_sampler():
            return ascend_sampler.forward(logits.clone(), sampling_metadata)

        for _ in range(NUM_WARMUP):
            run_ascend_sampler()
        torch.npu.synchronize()
        times = []
        for _ in range(NUM_ITERS):
            start_evt.record()
            run_ascend_sampler()
            end_evt.record()
            torch.npu.synchronize()
            times.append(start_evt.elapsed_time(end_evt))
        t_ascend = sum(times) / NUM_ITERS

        def run_gumbel():
            gumbel_sample(logits, idx_mapping, temperature, seed, pos, apply_temperature=True)

        for _ in range(NUM_WARMUP):
            run_gumbel()
        torch.npu.synchronize()
        times = []
        for _ in range(NUM_ITERS):
            start_evt.record()
            run_gumbel()
            end_evt.record()
            torch.npu.synchronize()
            times.append(start_evt.elapsed_time(end_evt))
        t_gumbel = sum(times) / NUM_ITERS

        speedup = t_gumbel / t_ascend if t_ascend > 0 else 0
        print(f"num_reqs={num_reqs}: AscendSampler {t_ascend:.3f} ms | v2 gumbel {t_gumbel:.3f} ms | speedup {speedup:.1f}x")

    print("=" * 60)


if __name__ == "__main__":
    main()
```

## Related

- v2 Model Runner: [Model Runner V2](https://github.com/vllm-project/vllm-ascend/issues/5208)
- AscendSampler: `vllm_ascend/sample/sampler.py`
- v2 Gumbel: `vllm_ascend/worker/v2/sample/gumbel.py`

### PR DESCRIPTION
**Pull request by lio1226**
### What this PR does / why we need it?

Fixes https://github.com/vllm-project/vllm-ascend/issues/7608.

In Model Runner V2, the Gumbel sampling kernel in `vllm_ascend/worker/v2/sample/gumbel.py` has excessive latency when `num_reqs` is large, and is significantly slower than `AscendSampler` in V1.

This PR replaces the Gumbel sampling Triton kernel with an Ascend C operator to reduce sampling overhead and improve serving throughput.

---

### Performance

> Lower is better for TTFT and TPOT. Higher is better for throughput.

#### Single-node serving

Configuration:

| Item | Value |
| --- | --- |
| Model | Qwen3-8B |
| Input length | 1K |
| Output length | 100 |
| Concurrency | 40 |
| Temperature | 0.01 |

| Method | TTFT | TPOT | Throughput |
| --- | ---: | ---: | ---: |
| Before | 2429.33 | 113.46 | 292.52 |
| This PR | 2230.41 | 77.08 | 405.38 |
| Improvement | -8.2% | -32.1% | +38.6% |

#### Multi-node serving

Configuration:

| Item | Value |
| --- | --- |
| Model | Qwen3-32B |
| Input length | 1K |
| Output length | 100 |
| Concurrency | 40 |

##### DP2 TP2

| Method | TTFT | TPOT | Throughput |
| --- | ---: | ---: | ---: |
| Before | 1685.70 | 142.83 | 245.56 |
| This PR | 1022.83 | 125.01 | 292.08 |
| Improvement | -39.3% | -12.5% | +18.9% |

##### PP2 TP2

| Method | TTFT | TPOT | Throughput |
| --- | ---: | ---: | ---: |
| Before | 2298.34 | 298.21 | 124.71 |
| This PR | 2153.71 | 238.99 | 153.80 |
| Improvement | -6.3% | -19.9% | +23.3% |

##### use speculative decoding when draft_sample_method=probabilistic
| Method | TTFT | TPOT | Throughput |
| --- | ---: | ---: | ---: |
| Before | 2929.477 | 65.460 | 417.203 |
| This PR | 2937.597 | 37.050 | 593.793 |
| Improvement | +0.28% | -43.40% | +42.33% |

---

### Accuracy

Accuracy was evaluated on Math500 with the following command:

```bash
ais_bench --models vllm_api_general_chat --datasets math500_gen_0_shot_cot_chat_prompt --debug
```

#### Temperature = 0.6

| Kernel | Dataset | Version | Metric | Mode | Accuracy |
| --- | --- | --- | --- | --- | ---: |
| Triton | math_prm800k_500 | f411c3 | accuracy | gen | 97.00 |
| Ascend C | math_prm800k_500 | f411c3 | accuracy | gen | 96.60 |

#### Temperature = 0

| Kernel | Dataset | Version | Metric | Mode | Accuracy |
| --- | --- | --- | --- | --- | ---: |
| Triton | math_prm800k_500 | f411c3 | accuracy | gen | 93.80 |
| Ascend C | math_prm800k_500 | f411c3 | accuracy | gen | 95.00 |

---

### Does this PR introduce _any_ user-facing change?

None.

---

### How was this patch tested?

The changes were verified with new E2E tests:

- `tests/e2e/nightly/single_node/ops/singlecard_ops/test_gumbel_sample.py`

Test environment:

- vLLM version: v0.26.0
- vLLM main: https://github.com/vllm-project/vllm/commit/0351e9aa1fdf1a51329d1906881528dfe61fc88e

### CODE REVIEW

**Comment by Ronald1995** [comment]
thanks for your effort. gumble sampling is equivalent to exponential sampling in v1, which will play a important role in precision, we can't evaluate the precision in single operator mode easily, please show the score results of math500 dataset to show ascendc implementation don't make a precision issue.

**Comment by lio1226** [comment]
> thanks for your effort. gumble sampling is equivalent to exponential sampling in v1, which will play a important role in precision, we can't evaluate the precision in single operator mode easily, please show the score results of math500 dataset to show ascendc implementation don't make a precision issue.

Hi @Ronald1995 , I have completed the accuracy test on the Math500 dataset, with details as follows.
`ais_bench --models vllm_api_general_chat --datasets math500_gen_0_shot_cot_chat_prompt --debug`
#### temp=0.6
triton
| dataset | version | metric | mode | vllm-api-general-chat |
|----- | ----- | ----- | ----- | -----|
| math_prm800k_500 | f411c3 | accuracy | gen | 97.00 |

ascendC
| dataset | version | metric | mode | vllm-api-general-chat |
|----- | ----- | ----- | ----- | -----|
| math_prm800k_500 | f411c3 | accuracy | gen | 96.60 |

In addition, I noticed that the current csrc directory has been refactored. Which directory should newly added custom operators be placed in the future?

**Review by ZT-AIA — DISMISSED** [review]

**Comment by Liamup777** [comment]
please squash your commits

**Comment by lio1226** [comment]
> please squash your commits

I have squashed the commits into a single one.

**Review by chenlongxiao — APPROVED** [review]
